File: core/src/main/java/javabot/model/Karma.java
package javabot.model;

import com.antwerkz.maven.SPI;
import org.bson.types.ObjectId;
import org.mongodb.morphia.annotations.Entity;
import org.mongodb.morphia.annotations.Id;
import org.mongodb.morphia.annotations.Index;
import org.mongodb.morphia.annotations.Indexes;
import org.mongodb.morphia.annotations.PrePersist;

import java.io.Serializable;
import java.time.LocalDateTime;

@Entity(value = "karma", noClassnameStored = true)
@SPI(Persistent.class)
@Indexes({
             @Index("upperName")
         })
public class Karma implements Serializable, Persistent {
    @Id
    private ObjectId id;

    private String name;

    private String upperName;

    private Integer value = 0;

    private String userName;

    private LocalDateTime updated;

    public ObjectId getId() {
        return id;
    }

    public void setId(ObjectId karmaId) {
        id = karmaId;
    }

    public String getName() {
        return name;
    }

    public void setName(String karmaName) {
        name = karmaName;
    }

    public Integer getValue() {
        return value;
    }

    public void setValue(Integer karmaValue) {
        value = karmaValue;
    }

    public String getUserName() {
        return userName;
    }

    public void setUserName(String usrName) {
        userName = usrName;
    }

    public LocalDateTime getUpdated() {
        return updated;
    }

    public void setUpdated(LocalDateTime date) {
        updated = date;
    }

    @PrePersist
    public void uppers() {
        upperName = name.toUpperCase();
    }
}

File: src/main/kotlin/javabot/model/Karma.kt
package javabot.model

import org.bson.types.ObjectId
import dev.morphia.annotations.Entity
import dev.morphia.annotations.Field
import dev.morphia.annotations.Id
import dev.morphia.annotations.Index
import dev.morphia.annotations.Indexes
import dev.morphia.annotations.PrePersist
import java.io.Serializable
import java.time.LocalDateTime

@Entity(value = "karma", useDiscriminator = false)
@Indexes(Index(fields = arrayOf(Field("upperName"))),
        Index(fields = arrayOf(Field("value"), Field("name"))))
class Karma : Serializable, Persistent {
    @Id
    var id: ObjectId? = null

    lateinit var name: String

    var upperName: String? = null

    var value: Int = 0

    lateinit var userName: String

    var updated: LocalDateTime? = null

    constructor() {
    }

    constructor(name: String, value: Int, userName: String) {
        this.name = name
        this.value = value
        this.userName = userName
    }

    @PrePersist fun uppers() {
        upperName = name.toUpperCase()
    }
}
